### ISSUE TITLE
Possible flaky JerseySupportTest

### ISSUE THREAD

**Issue by danielkec** [oracle/helidon#1161]
## Environment Details
* Helidon Version: SNAPSHOT
* Werker box: maven:3.6.1-jdk-11
----------

## Problem Description
Flaky test JerseySupportTest, happened to me already twice, once locally and [once on werker](https://app.wercker.com/Helidon/helidon/runs/build/5dd54f038e18590008b0f1cd?step=5dd54f2df947d100092c2cd1), cannot reproduce on re-run.
[WerkerHelidonFlakyTest.log](https://github.com/oracle/helidon/files/3876413/WerkerHelidonFlakyTest.log)

```
Running io.helidon.webserver.jersey.JerseySupportTest
Tests run: 28, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 4.365 sec <<< FAILURE! - in io.helidon.webserver.jersey.JerseySupportTest
nonJerseyGetNotFound  Time elapsed: 0.031 sec  <<< FAILURE!
org.opentest4j.AssertionFailedError: Unexpected error: 500 ==> expected: <SUCCESSFUL> but was: <SERVER_ERROR>
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:52)
	at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:186)
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:497)
	at io.helidon.webserver.jersey.JerseySupportTest.doAssert(JerseySupportTest.java:397)
	at io.helidon.webserver.jersey.JerseySupportTest.nonJerseyGetNotFound(JerseySupportTest.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:436)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:170)
	at org.junit.jupiter.engine.execution.ThrowableCollector.execute(ThrowableCollector.java:40)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:166)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:113)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:58)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:112)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:120)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:120)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:120)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:120)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:55)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:43)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at org.junit.platform.surefire.provider.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:132)
	at org.junit.platform.surefire.provider.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:111)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)


Results :

Failed tests: 
  JerseySupportTest.nonJerseyGetNotFound:255->doAssert:397 Unexpected error: 500 ==> expected: <SUCCESSFUL> but was: <SERVER_ERROR>

Tests run: 28, Failures: 1, Errors: 0, Skipped: 0

```

**Comment by barchetta** [comment]
Hit it today on the 1.4 branch. Here is the stdout:
```
[2019.12.12 13:25:57 FINE io.helidon.webserver.ForwardingHandler Thread[nioEventLoopGroup-3-1,10,main]: [Handler: 808147929] Received object: class io.netty.handler.codec.http.DefaultHttpRequest
2019.12.12 13:25:57 FINE io.helidon.webserver.ForwardingHandler Thread[nioEventLoopGroup-3-1,10,main]: [Handler: 808147929] Received object: class io.netty.handler.codec.http.LastHttpContent$1
2019.12.12 13:25:57 WARNING io.helidon.webserver.RequestRouting Thread[helidon-2,5,server]: Default error handler: Unhandled exception encountered.
java.util.concurrent.ExecutionException: Unhandled 'cause' of this exception encountered.
	at io.helidon.webserver.RequestRouting$RoutedRequest.defaultHandler(RequestRouting.java:377)
	at io.helidon.webserver.RequestRouting$RoutedRequest.nextNoCheck(RequestRouting.java:358)
	at io.helidon.webserver.RequestRouting$RoutedRequest.next(RequestRouting.java:399)
	at io.helidon.webserver.jersey.JerseyExampleMain.lambda$null$3(JerseyExampleMain.java:120)
	at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:870)
	at java.util.concurrent.CompletableFuture.uniExceptionallyStage(CompletableFuture.java:884)
	at java.util.concurrent.CompletableFuture.exceptionally(CompletableFuture.java:2196)
	at java.util.concurrent.CompletableFuture.exceptionally(CompletableFuture.java:110)
	at io.helidon.webserver.jersey.JerseyExampleMain.lambda$webServer$4(JerseyExampleMain.java:119)
	at io.helidon.webserver.RequestRouting$RoutedRequest.next(RequestRouting.java:310)
	at io.helidon.webserver.jersey.ResponseWriter.lambda$writeResponseStatusAndHeaders$0(ResponseWriter.java:118)
	at java.util.concurrent.CompletableFuture.uniRun(CompletableFuture.java:705)
	at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:687)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at io.helidon.webserver.jersey.JerseySupport$JerseyHandler.lambda$null$3(JerseySupport.java:267)
	at io.helidon.common.context.Contexts.runInContext(Contexts.java:98)
	at io.helidon.common.context.ContextAwareExecutorImpl.lambda$wrap$1(ContextAwareExecutorImpl.java:129)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.CompletionException: java.lang.IllegalStateException: Not enough request to submit item
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:647)
	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:669)
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:1997)
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:110)
	at io.helidon.webserver.jersey.JerseyExampleMain.lambda$webServer$4(JerseyExampleMain.java:111)
	... 12 more
Caused by: java.lang.IllegalStateException: Not enough request to submit item
	at io.helidon.common.reactive.BaseProcessor.submit(BaseProcessor.java:126)
	at io.helidon.common.reactive.MultiCollectingProcessor.hookOnComplete(MultiCollectingProcessor.java:46)
	at io.helidon.common.reactive.BaseProcessor.tryComplete(BaseProcessor.java:191)
	at io.helidon.common.reactive.BaseProcessor.subscribe(BaseProcessor.java:104)
	at io.helidon.common.reactive.Single.map(Single.java:46)
	at io.helidon.media.common.ContentReaders.readString(ContentReaders.java:66)
	at io.helidon.media.common.ContentReaders.lambda$stringReader$0(ContentReaders.java:89)
	at io.helidon.webserver.Request$InternalReader.apply(Request.java:220)
	at io.helidon.webserver.Request$Content.as(Request.java:297)
	at io.helidon.webserver.jersey.JerseyExampleMain.lambda$webServer$4(JerseyExampleMain.java:110)
	... 12 more
```

**Comment by romain-grecourt** [comment]
The code executed here is a very simple:

```java
request.content.as(String.class).thenAccept(...)
```

The reactive pipeline setup is equivalent to this:
```
CompletionStage<String> stage = Multi.from(chunks)
    .collect(new Collector<DataChunk, byte[]>() {
        final ByteArrayOutputStream baos = new ByteArrayOutputStream();

        @Override
        public void collect(DataChunk chunk) {
            try {
                Utils.write(chunk.data(), baos);
            } catch (IOException e) {
                throw new IllegalArgumentException("Cannot convert byte buffer to a byte array!", e);
            } finally {
                chunk.release();
            }
        }

        @Override
        public byte[] value() {
            return baos.toByteArray();
        }
    })
    .map((bytes) -> new String(bytes, StandardCharsets.UTF_8))
    .toStage();
```

This part of the stack trace is interesting:

```java
at io.helidon.common.reactive.MultiCollectingProcessor.hookOnComplete(MultiCollectingProcessor.java:46)
at io.helidon.common.reactive.BaseProcessor.tryComplete(BaseProcessor.java:191)
at io.helidon.common.reactive.BaseProcessor.subscribe(BaseProcessor.java:104)
```

It seems that the `MultiCollectingProcessor` instance was marked as done when the `SingleMappingProcessor` subscribed to it.

This means that `onComplete` was invoked instantly on the `MultiCollectingProcessor` instance after it subscribed to the upstream chunks. This is very likely the real culprit, yet I'm not sure under what circumstances would the webserver do that.

Also, `MultiCollectingProcessor` should be enhanced to not submit items downstream when no upstream items have been collected. It should be relatively easy to write a unit test for that scenario.

**Comment by romain-grecourt** [comment]
The bug in MultiCollectingProcessor can be easily reproduced by changing `Multi.from(chunks)` with `Multi.<DataChunk>empty()` in the snippet above.

**Comment by barchetta** [comment]
If you change the failing test to this, it fails most of the time:
```
    @Test
    public void nonJerseyGetNotFound() throws Exception {

        for (int n = 0; n < 1000; n++) {
            Response response = get("jersey/second");

            doAssert(response, "second-content: ");
        }
    }
```

**Comment by romain-grecourt** [comment]
If you replace `Multi.from(chunks)` with the following, it emits an empty string:
```
new Publisher<DataChunk>(){
    @Override
    public void subscribe(Flow.Subscriber<? super DataChunk> s) {
        s.onSubscribe(new Flow.Subscription() {
            @Override
            public void request(long l) {
                s.onComplete();
            }

            @Override
            public void cancel() {
            }
        });
    }
})
```

**Comment by danielkec** [comment]
Is it really a bad thing for `MultiCollectingProcessor` to do? I would expect collector to return empty collection when nothing came from upstream.
Java streams behaves in same manner:
```
Stream.empty().collect(Collectors.toList()).size() == 0
```
It would be nice to have back-pressure strategy with buffering for non-linear processors to avoid `Not enough request to submit item`. In case of `MultiCollectingProcessor` only small buffer is needed :)

**Comment by romain-grecourt** [comment]
It makes sense to align `Multi` with `java.util.Stream` as much as possible. This means that `Multi.collect` will return a Single of the collector container regardless of number of items published by a `Multi`.

This also preserves the current behavior for consuming an empty payload: `request.content().as(String.class)` completes with an empty String rather than exceptionally.

I guess the single item buffering should be implemented in `MultiCollectingProcessor` rather than directly in `BaseProcessor`.

I've also had in mind a base processor supporting back pressure, in fact I discussed this with @tomas-langer  last week ;). FYI I wrote a fancier processor based on OriginThreadPublisher when I worked on multipart, see [MultiPartDecoder](https://github.com/romain-grecourt/helidon/blob/multipart-take2/media/multipart/common/src/main/java/io/helidon/media/multipart/common/MultiPartDecoder.java) ; this could be a starting point for a back pressure base processor.
